List params are not expanded in native queries

The query below is not expanded into for the list and an exception is thrown
```
  @Query(value = "select t.* from hcp_tenant t " +
            "join hcp_tenant_consumer tc on tc.tenant_id = t.id and tc.topic_id in :topicIds",
            nativeQuery = true)
    Set<HcpTenantDTO> findByConsumersTopicsIdInList(List<Long> topicIds);
```

I have tried `(:topicIds)` as well but same result as the expansion isn't happening

```
SELECT hcp_topic_.`id`,hcp_topic_.`name`,hcp_topic_.`partitions`,hcp_topic_.`config`,hcp_topic_.`email`,hcp_topic_.`ms_id`,hcp_topic_.`compatibility`,hcp_topic_.`aha_id` FROM `hcp_topic` hcp_topic_ INNER JOIN `hcp_org` hcp_topic_hcp_org_ ON hcp_topic_.`hcp_org_id`=hcp_topic_hcp_org_.`id` INNER JOIN `hcp_env` hcp_topic_hcp_org_env_ ON hcp_topic_hcp_org_.`env_id`=hcp_topic_hcp_org_env_.`id` WHERE (hcp_topic_hcp_org_env_.`id` = ?)
10:09:13.254 [pool-2-thread-7] DEBUG io.micronaut.data.query - Executing Query: select t.* from hcp_tenant t join hcp_tenant_consumer tc on tc.tenant_id = t.id and tc.topic_id in ?
10:09:13.290 [pool-2-thread-7] ERROR i.m.h.s.netty.RoutingInBoundHandler - Unexpected error occurred: Unable to set PreparedStatement value: Parameter index out of range (2 > number of parameters, which is 1).
io.micronaut.data.exceptions.DataAccessException: Unable to set PreparedStatement value: Parameter index out of range (2 > number of parameters, which is 1).
	at io.micronaut.data.jdbc.mapper.JdbcQueryStatement.newDataAccessException(JdbcQueryStatement.java:273)
	at io.micronaut.data.jdbc.mapper.JdbcQueryStatement.setLong(JdbcQueryStatement.java:139)
	at io.micronaut.data.jdbc.mapper.JdbcQueryStatement.setLong(JdbcQueryStatement.java:33)
	at io.micronaut.data.runtime.mapper.QueryStatement.setDynamic(QueryStatement.java:134)
	at io.micronaut.data.jdbc.mapper.JdbcQueryStatement.setDynamic(JdbcQueryStatement.java:92)
	at io.micronaut.data.jdbc.mapper.JdbcQueryStatement.setDynamic(JdbcQueryStatement.java:33)
...

Caused by: java.sql.SQLException: Parameter index out of range (2 > number of parameters, which is 1).
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:129)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:89)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:63)
	at com.mysql.cj.jdbc.ClientPreparedStatement.checkBounds(ClientPreparedStatement.java:1402)
	at com.mysql.cj.jdbc.ClientPreparedStatement.getCoreParameterIndex(ClientPreparedStatement.java:1415)
	at com.mysql.cj.jdbc.ClientPreparedStatement.setLong(ClientPreparedStatement.java:1610)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.setLong(HikariProxyPreparedStatement.java)
	at io.micronaut.data.jdbc.mapper.JdbcQueryStatement.setLong(JdbcQueryStatement.java:137)
	... 127 common frames omitted


```

--- a/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2RepositorySpec.groovy
+++ b/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2RepositorySpec.groovy
@@ -271,6 +271,25 @@ class H2RepositorySpec extends AbstractRepositorySpec implements H2TestPropertyP
         cleanupData()
     }
 
+    void "test In Native Query function"() {
+        given:
+        savePersons(["Cemo", "Deniz", "Utku"])
+
+        when:"using a mix of parameters with collection types with IN queries"
+        def persons = pr.queryNames(
+            ["Ali"],
+            "James",
+            ["Onur"],
+            ["Cemo","Deniz","Olcay"]);
+
+        then:"The result is correct"
+        persons != null
+        persons.size() == 2
+
+        then:
+        cleanupData()
+    }
+
     void "test custom alias"() {
         given:
         saveSampleBooks()

--- a/data-model/src/main/java/io/micronaut/data/model/query/builder/QueryBuilder.java
+++ b/data-model/src/main/java/io/micronaut/data/model/query/builder/QueryBuilder.java
@@ -49,6 +49,11 @@ public interface QueryBuilder {
     Pattern VARIABLE_PATTERN = Pattern.compile("(:([a-zA-Z0-9]+))");
 
     /**
+     * A pattern used to find in queries in a query string.
+     */
+    Pattern IN_VARIABLES_PATTERN = Pattern.compile("(?<singleGroup>:[a-zA-Z0-9]+)|(?<inGroup>IN\\((:[a-zA-Z0-9]+)\\))");
+
+    /**
      * Builds an insert statement for the given entity.
      * @param repositoryMetadata The repository annotation metadata
      * @param entity The entity

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/intercept/AbstractQueryInterceptor.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/intercept/AbstractQueryInterceptor.java
@@ -56,6 +56,7 @@ import java.util.regex.Matcher;
 import java.util.stream.Collectors;
 
 import static io.micronaut.data.intercept.annotation.DataMethod.META_MEMBER_PAGE_SIZE;
+import static io.micronaut.data.model.query.builder.QueryBuilder.IN_VARIABLES_PATTERN;
 import static io.micronaut.data.model.query.builder.QueryBuilder.VARIABLE_PATTERN;
 
 /**
@@ -808,11 +809,11 @@ public abstract class AbstractQueryInterceptor<T, R> implements DataInterceptor<
         private final boolean isNumericPlaceHolder;
         private final boolean hasPageable;
         private final AnnotationMetadata annotationMetadata;
-        private final boolean hasIn;
         private final boolean isCount;
         private final DataType[] indexedDataTypes;
         private final String[] parameterNames;
         private final boolean hasResultConsumer;
+        private boolean hasIn;
         private Map<String, Object> queryHints;
         private Set<JoinPath> joinFetchPaths = null;
 
@@ -843,6 +844,17 @@ public abstract class AbstractQueryInterceptor<T, R> implements DataInterceptor<
             this.hasPageable = method.stringValue(PREDATOR_ANN_NAME, TypeRole.PAGEABLE).isPresent() ||
                                     method.stringValue(PREDATOR_ANN_NAME, TypeRole.SORT).isPresent() ||
                                     method.intValue(PREDATOR_ANN_NAME, META_MEMBER_PAGE_SIZE).orElse(-1) > -1;
+
+            Matcher inMatcher = IN_VARIABLES_PATTERN.matcher(query);
+            StringBuffer sb = new StringBuffer();
+            for (int i = 1; inMatcher.find(); i++) {
+                if (inMatcher.group("inGroup") != null) {
+                    inMatcher.appendReplacement(sb, "\\?\\$IN("+ i + ")");
+                    this.hasIn = true;
+                }
+            }
+            inMatcher.appendTail(sb);
+            query = sb.toString();
 
             if (isNumericPlaceHolder && method.isTrue(Query.class, DataMethod.META_MEMBER_RAW_QUERY)) {
                 Matcher matcher = VARIABLE_PATTERN.matcher(query);

--- a/data-tck/src/main/java/io/micronaut/data/tck/repositories/PersonRepository.java
+++ b/data-tck/src/main/java/io/micronaut/data/tck/repositories/PersonRepository.java
@@ -58,6 +58,9 @@ public interface PersonRepository extends CrudRepository<Person, Long>, Pageable
     @Query("UPDATE person SET name = 'test' WHERE id = :id")
     long updatePersonCustom(Long id);
 
+    @Query("SELECT * FROM person WHERE name = :names1 or name IN(:names3) or name IN(:names0)")
+    List<Person> queryNames(List<String> names0, String names1, List<String> names2, List<String> names3);
+
     CompletableFuture<Long> updatePersonFuture(@Id Long id, @Parameter("name") String name);
 
     long updateByName(String name, int age);

